fppanalysis/time_delay_estimation.py

```diff
@@ -0,0 +1,153 @@
+from scipy.stats import gamma
+import correlation_function as cf
+
+
+def estimate_delays(x, y, dt, distribution=gamma, ax=None, plot="full", min_cutoff=0):
+    """
+    Use:
+       estimate_delays(x, y, dt, distribution=gamma, ax=None, plot="full", min_cutoff=0)
+
+    Estimates the time delay distribution parameters for the pulses propagating from two measurement points.
+    This is done by optimizing the time delay distribution parameters such that the predicted cross-correlation,
+    given by the convolution of the autocorrelation and the time delay distribution, best fits the observed cross-
+    correlation.
+
+    Time series x, y, corresponding to each point measurement must be provided. The time delay distribution
+    is assumed to follow a family distribution given by the argument distribution. Plots relevant autocorrelation
+    and cross-correlation functions if a matplotlib ax is provided.
+
+    Input:
+        x: Time series ...................... (N,) np array
+        x: Time series ...................... (N,) np array
+        distribution: Assumed distribution .... class implementing scipy.stats.rv_continuous
+        ax: Optional, if a matplotlib.pyplot.axis is provided, relevant plots will be plotted. These are meant to help
+        understand the underlying principles of the optimization, the plots are not suitable for scientific publication
+        .
+        plot: String, if "full", plots more stuff.
+        min_cutoff: An upper bound for the cross-correlation maxima. Usage: If working with short time series
+        or noisy data, it can be helpful for the method to set this value to an upper bound for the time that
+        maximizes the cross-correlation, that is, a time such that you are sure that the cross-correlation is
+        maximized before it.
+    Output:
+        avg: Average delay time
+        arg: Possible distribution shape parameter. None if distribution is shapeless
+        scale: Distribution scale parameter.
+    """
+    import numpy as np
+    from scipy.stats import uniform, norm
+    from scipy.signal import fftconvolve
+    from scipy.optimize import minimize
+
+    def get_pdf(params, tauR):
```

Comment:
Maybe I am pedantic, but wouldn't it be more readable if the functions don't use any variables outside of the function scope (adding "distributions" as a parameter and defining the function outside of estimate_delays)?

Reply:
Updated. I left the get_error function as it was though, as it is too short and would have too many arguments. If python had the syntax for it I would have made a lambda function out of it instead, but this is not possible in python.